# 딥러닝 기초

## CH3.3~CH3.7 과제

## 2020100381 안정빈

## 3-3. 다차원 배열의 계산

다차원 배열: '숫자의 집합'으로, 숫자를 N차원으로 나열하는 것을 뜻함.

In [33]:
# 원소가 4개인 1차원 배열 

import numpy as np 
A=np.array([1,2,3,4]) # 1차원 배열
print(A)
print("배열의 차원:",np.ndim(A)) # np.ndim 배열의 차원 수를 확인하는 함수 
print("배열의 형상:",A.shape) # 배열의 형상 확인
print("원소의 개수:",A.shape[0]) # 첫번째 축의 크기 출력/ 즉 1차원 배열에선 원소의 개수

[1 2 3 4]
배열의 차원: 1
배열의 형상: (4,)
원소의 개수: 4


In [34]:
# 2차원 배열 

B=np.array([[1,2],[3,4],[5,6]]) # 3X2 배열 생성 
print(B)
print("배열의 차원:",np.ndim(B))
print("배열의 형상:",B.shape)

[[1 2]
 [3 4]
 [5 6]]
배열의 차원: 2
배열의 형상: (3, 2)


2 차원 배열은 행렬이라 부르며, 배열의 가로 방향을 행, 세로 방향을 열이라고 부름

-> 위에 출력 결과에서 1,2는 행/ 1,3,5는 열이라고 부름 

### 3-3-2. 행렬의 곱

A=[[1,2],[3,4]] / B= [[5,6],[7,8]] A와 B, 즉 2x2 행렬의 곱은, 왼쪽 행렬의 행과 오른쪽 행렬의 열을 원소별로 곱하고 그 값들을 더해서 계산

A의 1행 X B의 1열 => 1행 1번째 원소 1x5 + 2x7 = 19

A의 1행 X B의 2열 => 1행 2번째 원소 1x6 + 2x8 = 22

A의 2행 X B의 1열 => 2행 1번째 원소 3x5 + 4x7 = 43

A의 2행 X B의 2열 => 2행 2번째 원소 3x6 + 4x8 = 50

In [35]:
A=np.array([[1,2],[3,4]])
print("A의 형상:",A.shape)
B=np.array([[5,6,],[7,8]])
print("B의 형상:",B.shape)
np.dot(A,B) # 행렬의 곱/ np.dot은 입력이 1차원 배열이면 벡터, 2차원 배열이면 행렬 곱을 계산


A의 형상: (2, 2)
B의 형상: (2, 2)


array([[19, 22],
       [43, 50]])

In [36]:
A=np.array([[1,2,3],[4,5,6]]) # 2x3 배열
B=np.array([[1,2],[3,4],[5,6]]) # 3x2 배열 

print("A의 형상:",A.shape)
print("B의 형상:",B.shape)
np.dot(A,B)

A의 형상: (2, 3)
B의 형상: (3, 2)


array([[22, 28],
       [49, 64]])

In [37]:
C=np.array([[1,2],[3,4]])
print("C의 형상:",C.shape)
print("A의 형상:",A.shape)
# np.dot(A,C) 
# 배열의 형상 (A의 1번째 차원의 원소 수와 C의 0번째 차원의 원소 수)가 다르므로 오류 발생

C의 형상: (2, 2)
A의 형상: (2, 3)


In [38]:
A=np.array([[1,2],[3,4],[5,6]])
B=np.array([7,8])


print("A의 형상:",A.shape)
print("B의 형상:",B.shape)
np.dot(A,B)

A의 형상: (3, 2)
B의 형상: (2,)


array([23, 53, 83])

행렬의 곱은, 행렬 A의 1번째 차원의 원소 수(열의 개수)와 행렬 B의 0번째 차원의 원소 수(행의 개수)가 같아야 한다.

즉, 2X3 행렬 A와, 2X2 행렬 C의 곱은 (A의 1번째 차원의 원소 수는 3개, C의 0번째 차원의 원소 수 2개)이므로 불가하다.

A는 5x2 행열, B는 2X4 행열이라면 원소 수(2)가 일치하므로, 5X4 행렬이 만들어진다.

B가 1차원 배열일 때도, 대응하는 차원의 원소 수가 같아야 한다.

결론: 행렬의 곱에서는 대응하는 차원의 원소 수가 일치해야 하며, 일치할 경우, A의 0번째 원소 수(행 수)와 B의 1번째 원소 수(열 수)의 크기인 행렬이 만들어진다.

### 3-3-3. 신경망에서의 행렬 곱

In [39]:
X=np.array([1,2])
W=np.array([[1,3,5],[2,4,6]])

print("X의 형상:",X.shape)
print("W(가중치)의 형상:",W.shape)
Y=np.dot(X,W)
print(Y)

X의 형상: (2,)
W(가중치)의 형상: (2, 3)
[ 5 11 17]


신경망에서의 계산을 행렬 계산으로 정리 가능

배열을 사용하면 계산을 빠르고 효율적으로 할 수 있으며, 복잡한 연산 처리 가능하다.

또한 데이터 및 가중치, 편향을 배열로써 관리가 가능하다.


## 3-4. 3층 신경망 구현하기

입력층 2개, 첫 번째 은닉층 3개, 두 번째 은닉층 2개, 출력층 2개인 신경망 구현

In [40]:
# 입력층에서 1층으로 신호 전달
X=np.array([1.0,0.5]) # 입력값
W1=np.array([[0.1,0.3,0.5],[0.2,0.4,0.6]]) # 가중치
B1=np.array([0.1,0.2,0.3]) # 편향

print("가중치의 형상:",W1.shape)
print("입력값의 형상:",X.shape)
print("가중치의 형샹:",B1.shape)

A1=np.dot(X,W1)+B1 # 입력값과 가중치를 곱한 후, 편향을 더 함.

가중치의 형상: (2, 3)
입력값의 형상: (2,)
가중치의 형샹: (3,)


In [41]:
# 활성화 함수인 시그모이드를 이용하여, 첫번째 은닉층의 출력값 계산
def sigmoid(x):
    return 1/(1+np.exp(-x))

Z1=sigmoid(A1)
print(A1) # 선형 변환된 값
print(Z1) # 첫 번째 은닉층의 출력값

[0.3 0.7 1.1]
[0.57444252 0.66818777 0.75026011]


In [42]:
# 첫번째 은닉층에서 두번째 은닉층으로의 신호 전달
W2=np.array([[0.1,0.4],[0.2,0.5],[0.3,0.6]])
B2=np.array([0.1,0.2])

print("첫 번째 은닉층의 출력값 형상:", Z1.shape)
print("2 번째 가중치 형상:",W2.shape)
print("편향의 형상:",B2.shape)

A2=np.dot(Z1, W2)+B2 # 두 번째 은닉층의 입력값
Z2=sigmoid(A2) # 두 번째 은닉층의 출력 값

첫 번째 은닉층의 출력값 형상: (3,)
2 번째 가중치 형상: (3, 2)
편향의 형상: (2,)


In [43]:
# 두번째 은닉층에서 출력층으로의 신호 전달 

def identity_function(x): # 활성화 함수로 항등 함수 사용/ 항등 함수는 입력을 그대로 출력
    return x

W3=np.array([[0.1,0.3],[0.2,0.4]])
B3=np.array([0.1,0.2])

A3=np.dot(Z2,W3)+B3
Y=identity_function(A3) # y=A3

출력층의 활성화 함수는 문제의 성질에 맞게 사용

항등 함수: 회귀

시그모이드 함수: 이진 분류 

소프트맥스 함수: 다중 분류 

### 3-4-3. 구현 정리

In [44]:
def init_network(): # 가중치와 편향을 초기화, 딕셔너리 변수에 저장
    network={}
    network['W1']=np.array([[0.1,0.3,0.5],[0.2,0.4,0.6]])
    network['b1']=np.array([0.1,0.2,0.3])
    network['W2']=np.array([[0.1,0.4],[0.2,0.5],[0.3,0.6]])
    network['b2']=np.array([0.1,0.2])
    network['W3']=np.array([[0.1,0.3],[0.2,0.4]])
    network['b3']=np.array([0.1,0.2])

    return network

def forward(network,x): # 입력 신호를 출력으로 변환하는 처리 과정 구현 / 입력->출력 즉 순전파 구현
    W1, W2, W3 = network['W1'], network['W2'], network['W3']
    b1, b2, b3= network['b1'], network['b2'], network['b3']

    a1=np.dot(x,W1) + b1
    z1=sigmoid(a1)
    a2=np.dot(z1,W2) + b2
    z2=sigmoid(a2)
    a3=np.dot(z2,W3)+ b3
    y=identity_function(a3)

    return y

network=init_network()
x=np.array([1.0,0.5])
y=forward(network,x)
print(y)

[0.31682708 0.69627909]


## 3-5. 출력층 설계하기

### 3-5-1. 항등 함수와 소프트맥스 함수 구현하기 

신경망은 분류와 회귀 문제 모두 이용 가능, 그러나 분류와 회귀일 때, 출력층에서 사용하는 활성화 함수가 다름.

항등 함수: 회귀 / 소프트맥스 함수: 분류 문제

항등 함수: 입력을 그대로 출력하는 함수 -> 출력층에서 사용 시, 입력 신호가 그대로 출력 신호가 된다.

소프트맥스 함수: 입력 신호 ak의 지수 함수/ 모든 입력 신호의 지수 함수의 합

In [45]:
a=np.array([0.3,2.9,4.0])

exp_a=np.exp(a) # 지수 함수 
print(exp_a)

[ 1.34985881 18.17414537 54.59815003]


In [46]:
sum_exp_a=np.sum(exp_a) # 지수 함수의 합
print(sum_exp_a)

74.1221542101633


In [47]:
y= exp_a/ sum_exp_a 

print(y)

[0.01821127 0.24519181 0.73659691]


In [48]:
def softmax(a): # 소프트맥스 함수
    exp_a=np.exp(a)
    sum_exp_a=np.sum(exp_a)
    y= exp_a/ sum_exp_a

    return y
    

### 3-5-2. 소프트맥스 함수 구현 시 주의점

오버플로 문제: 컴퓨터는 수를 크기가 유한한 데이터로 다뤄, 수의 범위가 한정되어 너무 큰 값을 다룰 수 없음.

소프트맥스 함수는 지수 함수를 사용하는데, 지수 함수는 아주 큰 값을 만들기에, 이런 큰 값끼리 나눗셈을 하면 결과 수치가 '불안정'해짐

In [49]:
a=np.array([1010,1000,990])
np.exp(a)/np.sum(np.exp(a)) # 소프트맥스 함수의 계산

C:\Users\jeongbin\AppData\Local\Temp\ipykernel_23532\2501791746.py:2: RuntimeWarning: overflow encountered in exp
  np.exp(a)/np.sum(np.exp(a)) # 소프트맥스 함수의 계산
C:\Users\jeongbin\AppData\Local\Temp\ipykernel_23532\2501791746.py:2: RuntimeWarning: invalid value encountered in divide
  np.exp(a)/np.sum(np.exp(a)) # 소프트맥스 함수의 계산


array([nan, nan, nan])

In [50]:
c=np.max(a) 
a-c

array([  0, -10, -20])

In [51]:
np.exp(a-c)/ np.sum(np.exp(a-c))


array([9.99954600e-01, 4.53978686e-05, 2.06106005e-09])

In [52]:
def softmax(a):
    c=np.max(a)
    exp_a=np.exp(a-c)
    sum_exp_a=np.sum(exp_a)
    y= exp_a/ sum_exp_a

    return y

# 오버 플로를 막기 위해 입력 신호 중 최댓값을 빼서 계산

소프트맥스의 지수 함수를 계산할 때 어떤 정수를 더하거나 빼도 결과가 바뀌지 않음,

즉, 오버플로를 막기 위해 입력 신호 중 최댓값을 이용

### 3-5-3. 소프트맥스 함수의 특징

소프트맥스 함수의 출력은 0~1.0 사이의 실수이며, 소프트맥스 함수 출력의 총합은 1 => 소프트맥스 함수의 출력을 '확률'로 해석 가능

In [53]:
a=np.array([0.3, 2.9, 4.0])
y=softmax(a)
print(y) 
print("소프트맥스 함수의 합계:",np.sum(y))

[0.01821127 0.24519181 0.73659691]
소프트맥스 함수의 합계: 1.0


y[0]의 확률은 0.018, y[1]의 확률은 0.245, y[2]의 확률은 0.736으로 해석 가능

-> '2번째 원소의 확률이 가장 높으니, 답은 2번 클래스'라고 할 수 있음

즉, 소프트맥스 함수를 이용하여 문제를 확률적으로 대응 가능 

그러나, 소프트맥스 함수를 적용해도 각 원소의 대소 관계는 변하지 않는다. 그러므로 출력층의 소프트맥스 함수는 생략 가능

### 3-6. 손글씨 숫자 인식

MNIST 데이터셋: 0부터 9까지의 숫자 이미지로 구성된 데이터셋

In [54]:
import sys, os
sys.path.append(os.path.join(os.getcwd(), '..'))
from dataset.mnist import load_mnist 

(x_train, t_train),(x_test,t_test)= \
    load_mnist(flatten= True, normalize=False) 
# flatten: 입력 이미지를 1차원 배열로 만들지 여부 / normalize: 입력 이미지의 픽셀값을 0~1.0 사이로 정규화 시킬 것인지 여부

print(x_train.shape)
print(t_train.shape)
print(x_test.shape)
print(t_test.shape)

(60000, 784)
(60000,)
(10000, 784)
(10000,)


In [55]:
from PIL import Image # 이미지를 표시하는 모듈

def img_show(img):
    pil_img=Image.fromarray(np.uint8(img))
    pil_img.show()

img=x_train[0]
label=t_train[0]
print(label)

print("기존 형상:",img.shape)
img=img.reshape(28,28) # flatten=True로 하여 1차원으로 저장되었기에, 원래 형상인 28x28로 변형
print("변경된 형상:",img.shape)

img_show(img)

5
기존 형상: (784,)
변경된 형상: (28, 28)


### 3-6-2. 신경망의 추론 처리 

입력층 뉴런 784개로 구성 이미지의 크기가 28x28=784이기 때문.

출력층 뉴런 10개로 구성 0~9까지 총 10개의 숫자로 구분 / 출력층의 뉴런 개수는 풀려는 문제에 맞게 적절히 정해야 함.

In [56]:
import pickle

def get_data():
    (x_train, t_train), (x_test,t_test)= \
        load_mnist(normalize=True, flatten=True, one_hot_label=False) # normalize=True를 이용해 정규화 처리

    return x_test, t_test
    
file_path = "C:/Users/jeongbin/study/DL_study/dataset/sample_weight.pkl"

def init_network(): # 학습된 가중치 매개변수 읽음
     with open(file_path, 'rb') as f:
        network = pickle.load(f)
         
     return network
    
def predict(network, x):
    W1, W2, W3=network['W1'], network['W2'], network['W3']
    b1, b2, b3= network['b1'], network['b2'], network['b3']

    a1= np.dot(x,W1)+b1
    z1=sigmoid(a1)
    a2=np.dot(z1,W2)+b2
    z2=sigmoid(a2)
    a3=np.dot(z2,W3)+b3
    y=softmax(a3)

    return y

In [57]:
x, t= get_data()
network=init_network()

accuracy_cnt=0
for i in range(len(x)):
    y=predict(network, x[i])
    p=np.argmax(y) # 확률이 가장 높은 원소의 인덱스를 구해 예측 결과로 이용
    if p == t[i]:
        accuracy_cnt+=1 

print("Accuracy:"+str(float(accuracy_cnt) / len(x)))

Accuracy:0.9352


### 3-6-3. 배치 처리

배치: 하나로 묶은 입력 데이터

In [58]:
# 이미지를 한장씩 입력했을 때 처리 흐름
x,_=get_data()
network= init_network()
W1, W2, W3=network['W1'], network['W2'], network['W3']

print(x.shape)
print(x[0].shape)
print(W1.shape)
print(W2.shape)
print(W3.shape)


(10000, 784)
(784,)
(784, 50)
(50, 100)
(100, 10)


In [59]:
# 이미지를 한번에 백 장씩 입력 처리
x,_=get_data()
network= init_network()

batch_size= 100 # 배치 크기 
accuracy_cnt=0

for i in range(0,len(x), batch_size):
    x_batch = x[i:i+batch_size] #batch_size만큼, 100만큼 묶어서 꺼냄
    y_batch =predict(network, x_batch)
    p=np.argmax(y_batch, axis=1) #100x10 배열 중 1번째 차원을 축으로 최댓값의 인덱스를 찾도록 설정
    accuracy_cnt+=np.sum(p==t[i:i+batch_size])
    
print("Accuracy:"+str(float(accuracy_cnt) / len(x)))

Accuracy:0.9352


In [60]:
# range(start, end, step): start부터 end-1까지, step간격으로 증가
print(list(range(0,10)))
print(list(range(0,10,3)))

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
[0, 3, 6, 9]


In [61]:
x=np.array([[0.1,0.8,0.1],[0.3,0.1,0.6],
            [0.2,0.5,0.3],[0.8,0.1,0.1]])
y=np.argmax(x,axis=1)
print(y)

[1 2 1 0]


In [62]:
y=np.array([1,2,1,0])
t=np.array([1,2,0,0])

print(y==t)
print(np.sum(y==t))

[ True  True False  True]
3


데이터를 배치로 처리함으로써 효울적이고 빠른 처리가 가능하다.